Display RadioSchedules on Simorgh Frontpage

itimirichard opened the pull request:
Resolves #4671

**Overall change:** 
_Added Radio Schedules to Frontpage._

**Code changes:**
- Add RadioScheduleContainer to simorgh frontpage
- Ensure RadioSchedule is only displayed on Canonical pages
- Compose RadioSchedule component with the Section Label and RadioFrequencyLink
- Added unit tests and snapshots._

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] I have added labels to this PR for the relevant pod(s) affected by these changes
- [x] I have assigned this PR to the Simorgh project

**Testing:**
- [x] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`) 
- [x] This PR requires manual testing

RTL
![Screenshot 2020-03-12 at 13 02 34](https://user-images.githubusercontent.com/12303856/76510641-fa789300-6462-11ea-8404-5649014bdbd1.png)

itimirichard committed:
initial commit: move data fetching logic to canonical

itimirichard committed:
Merge remote-tracking branch 'origin/latest' into add-radioschedules-to-frontpage

itimirichard committed:
remove endpoint from frontpage

itimirichard committed:
added override endpoint for storybook

itimirichard committed:
Merge remote-tracking branch 'origin/latest' into add-radioschedules-to-frontpage

itimirichard committed:
put radio schedules after topstories slic

itimirichard committed:
Merge branch 'latest' into add-radioschedules-to-frontpage

itimirichard committed:
resolve conflict

Commit:
Add radio frequency link styling.

Commit:
Update renderFrequencyLink().

itimirichard committed:
refactor getRadioscheduleEndpoint

tochwill commented on src/app/containers/RadioSchedule/Canonical/index.jsx:
```diff
@@ -0,0 +1,87 @@
+const logger = webLogger();
+
+const CanonicalRadioSchedule = ({ endpoint }) => {
+  const header = 'BBC News Radio';
```
Could you add this as a translation for each service? It can be in english for now but we'll want to get it from the translation config :)

itimirichard replied:
This has been added in https://github.com/bbc/simorgh/pull/5812

tochwill commented on src/app/containers/RadioSchedule/Canonical/index.jsx:
```diff
@@ -0,0 +1,87 @@
+
+  const handleResponse = async response => {
+    const radioScheduleData = await response.json();
+    setRadioSchedule(radioScheduleData.schedules.slice(0, 4));
```
Cool, we'll update this based on Olga's PR!

tochwill commented on src/app/containers/RadioSchedule/Canonical/index.jsx:
```diff
@@ -0,0 +1,87 @@
+
+  const renderFrequencyLink = link => (
+    <RadioFrequencyLink link={link}>
+      Radio frequency link text
```
Could you also put these into the translation config?

itimirichard replied:
This has been added in https://github.com/bbc/simorgh/pull/5812

tochwill commented on src/app/containers/RadioSchedule/index.jsx:
```diff
@@ -1,68 +1,41 @@
+// The logic here is sufficient for now for testing locally. This should change once we are ready to put radioschedules on live
+const getRadioScheduleEndpoint = (service, environment) =>
+  environment === 'local'
+    ? `/${service}/bbc_${service}_radio/radioschedule.json`
```
These shouldn't be different, could we just update the local file names to match what's defined in mozart?

tochwill commented on src/app/containers/RadioSchedule/Canonical/index.jsx:
```diff
@@ -0,0 +1,87 @@
+    fetchRadioScheduleData(endpoint);
+  }, [endpoint]);
+
+  const RadioFrequencyLink = styled(Link)`
```
Could you create these components outside of the main CanonicalRadioSchedule component?

Bopchy replied:
Moved `RadioFrequencyLink` above CanonicalRadioSchedule

Commit:
Update local radio schedule data file names.

Commit:
Move RadioFrequencyLink styled component outside of CanonicalRadioSchedule.

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into add-radioschedules-to-frontpage

itimirichard committed:
Merged latest into branch

itimirichard committed:
pull in preprocessing logic for schedules

itimirichard committed:
remove bar from section label

Commit:
Fix failing tests.

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into add-radioschedules-to-frontpage

Commit:
Update bundle MAX_SIZE

itimirichard committed:
add margin wrapper for different breakpoints

itimirichard committed:
additional styling

itimirichard committed:
Merge branch 'add-radioschedules-to-frontpage' of https://github.com/bbc/simorgh into add-radioschedules-to-frontpage

Commit:
Add tests.

Commit:
Merge branch 'add-radioschedules-to-frontpage' of github.com:bbc/simorgh into add-radioschedules-to-frontpage

Commit:
Merge branch 'latest' into add-radioschedules-to-frontpage

Commit:
Add snapshots.

itimirichard committed:
pull in latest version of psammead-radio-schedule

itimirichard committed:
update snapshots

itimirichard committed:
update snapshots

itimirichard committed:
updated snapshots, refactor

Commit:
Update snapshot.

thekp commented on scripts/bundleSizeConfig.js:
```diff
@@ -2,5 +2,5 @@ module.exports = {
   // Size limit for all bundles used by each service (K)
   // Keep these +/- 5K and update frequently!
   MIN_SIZE: 702,
```
I just checked locally on your PR for the smallest bundled service atm, which is currently 724.
![image](https://user-images.githubusercontent.com/30599794/76513341-132a7e00-644e-11ea-87d1-75767cc21752.png)

so we should also update the min size here to be -5 of 724

EinsteinNjoroge commented on src/app/containers/RadioSchedule/Canonical/index.jsx:
```diff
@@ -0,0 +1,130 @@
+  );
+
+  return (
+    <FrontPageRadioScheduleSection>
```
Since this is the default section, we should find a more descriptive name for it.
Perhaps `ConstrainedRadioScheduleSection`

tochwill replied:
I can't see a reason why the styles can't just be added to `RadioScheduleSection`, so I think we can just do that?

itimirichard replied:
Thanks, the styles have been consolidated into `RadioScheduleSection`.

OlgaLyubin commented on src/app/containers/RadioSchedule/Canonical/index.jsx:
```diff
@@ -0,0 +1,130 @@
+  }
+`;
+
+const RadioFrequencyLink = styled(Link)`
```
Should this also have `cursor: pointer;`? It currently does not look like a link on mouse hover

tochwill replied:
Huh, `Link` is just an `a` tag. It should display a pointer if there is an href set

tochwill replied:
Ah, I've mentioned this below

thekp commented on src/app/containers/RadioSchedule/utilities/testHelpers.jsx:
```diff
@@ -0,0 +1,42 @@
+import { ServiceContextProvider } from '#contexts/ServiceContext';
+import { ToggleContext } from '#contexts/ToggleContext';
+
+const localRadioScheduleEndpoint = service => {
```
I don't think we need this function. Looks like we are only using it for our RadioSchedulesWithContext component below but that is only used in unit tests which will have mock fetched data.

Bopchy replied:
Removed `localRadioScheduleEndpoint()`

thekp replied:
Thanks Ruth 💯

tochwill commented on src/app/containers/RadioSchedule/Canonical/index.jsx:
```diff
@@ -0,0 +1,130 @@
+  'aria-labelledby': 'Radio-Schedule',
+}))``;
+
+const FrontPageRadioScheduleSection = styled(RadioScheduleSection)`
```
Could you just add this styling to the `RadioScheduleSection` above?

tochwill commented on src/app/containers/RadioSchedule/Canonical/index.jsx:
```diff
@@ -0,0 +1,130 @@
+  }
+
+  const renderFrequencyLink = (link, label) => (
+    <RadioFrequencyLink link={link}>{label}</RadioFrequencyLink>
```
I don't think the `link` prop is valid - `Link` is just a styled `a` tag, so you should be passing an `href` here :)

EinsteinNjoroge commented on src/app/containers/RadioSchedule/Canonical/index.jsx:
```diff
@@ -0,0 +1,130 @@
+          dir={dir}
+        />
+        {frequenciesPageUrl &&
+          renderFrequencyLink(frequenciesPageUrl, frequenciesPageLabel)}
```
```suggestion
          renderFrequencyLink()}
```

tochwill replied:
Does this even need to be a function? We could just have something like

```
{frequenciesPageUrl && (
  <RadioFrequencyLink href={frequenciesPageUrl}>
    {frequenciesPageLabel}
  </RadioFrequencyLink>
)}
```

itimirichard replied:
Thanks, this suggestion has been effected!

tochwill commented on src/app/containers/RadioSchedule/utilities/processRadioSchedule.js:
```diff
@@ -0,0 +1,69 @@
+        startTime: program.publishedTimeStart,
+        link: getLink(currentState, program, service),
+        brandTitle: program.brand.title,
+        episodeTitle: program.episode.presentationTitle,
```
Could we use the ramda `pathOr` function that we use throughout simorgh to access this deeply nested data, to avoid running into trouble if any of these are undefined?

itimirichard replied:
This has been done!

EinsteinNjoroge commented on src/app/containers/RadioSchedule/Canonical/index.jsx:
```diff
@@ -0,0 +1,130 @@
+
+  const renderFrequencyLink = (link, label) => (
+    <RadioFrequencyLink link={link}>{label}</RadioFrequencyLink>
+  );
```
```suggestion
const renderFrequencyLink = () => (
    <RadioFrequencyLink link={frequenciesPageUrl}>
        {frequenciesPageLabel}
    </RadioFrequencyLink>
);
```

EinsteinNjoroge commented on src/app/containers/RadioSchedule/Canonical/index.jsx:
```diff
@@ -0,0 +1,130 @@
+    fetchRadioScheduleData(endpoint);
+  }, [endpoint, service, script, timezone, locale]);
+
+  if (!schedule) {
+    return null;
+  }
```
We should retain this in the `RadioShedules/index.js` file

itimirichard replied:
@EinsteinNjoroge The `RadioShedules/index.js` file has no notion of `schedule` data, just the component (Canonical in this case) and logic to render it i.e. `hasRadioSchedule`, `isAmp`

itimirichard committed:
refactor RadioSchedule/Canonical

itimirichard committed:
refactor data processing to use pathOr

Commit:
Update bundleSize MIN_SIZE

Commit:
Remove localRadioScheduleEndpoint().

Commit:
Merge branch 'add-radioschedules-to-frontpage' of github.com:bbc/simorgh into add-radioschedules-to-frontpage

Commit:
Update bundle size.

thekp committed:
Merge branch 'latest' into add-radioschedules-to-frontpage

mmcculloch commented on src/app/containers/RadioSchedule/Canonical/index.jsx:
```diff
@@ -0,0 +1,127 @@
+    const fetchRadioScheduleData = pathname =>
+      fetch(pathname, { mode: 'no-cors' })
+        .then(handleResponse)
+        .catch(e => logger.error(`HTTP Error: "${e}"`));
```
We should double check what these errors look like in testing/locally.

mmcculloch commented on src/app/containers/RadioSchedule/Canonical/index.test.jsx:
```diff
@@ -0,0 +1,23 @@
+    await matchSnapshotAsync(
+      <RadioSchedulesWithContext service="hausa" radioScheduleToggle />,
+    );
+  });
```
Add some assertion tests for this component.

Bopchy replied:
These will be added in #5871

mmcculloch commented on src/app/containers/RadioSchedule/utilities/processRadioSchedule.js:
```diff
@@ -0,0 +1,68 @@
+    });
+
+  return schedules;
+};
```
Add unit tests for this file.

Bopchy replied:
These will be added in #5871

Commit:
Move tests.
- Move render related tests to RadioSchedule/Canonical/index.test.jsx from RadioSchedule/index.test.jsx
- Reomove hausa test as it also had page frequency URL

Commit:
Merge branch 'add-radioschedules-to-frontpage' of github.com:bbc/simorgh into add-radioschedules-to-frontpage

itimirichard committed:
Merge branch 'latest' into add-radioschedules-to-frontpage

thekp commented on src/app/containers/RadioSchedule/utilities/processRadioSchedule.js:
```diff
@@ -0,0 +1,68 @@
+import moment from 'moment';
+import { pathOr, findLastIndex, propSatisfies } from 'ramda/src';
```
quick change here. the bundle size has significantly increased since ramda has been imported. 

we should update this import so that we are importing like:
```
import pathOr from 'ramda/src/pathOr';
import findLastIndex from 'ramda/src/findLastIndex';
import propSatisfies from 'ramda/src/propSatisfies';
```

Commit:
Change how ramda functions are imported. Update snapshot. Reduce bundle size.

Commit:
Merge branch 'add-radioschedules-to-frontpage' of github.com:bbc/simorgh into add-radioschedules-to-frontpage

PriyaKR committed:
Merge branch 'latest' into add-radioschedules-to-frontpage

tochwill commented on src/app/containers/RadioSchedule/index.jsx:
```diff
@@ -1,128 +1,38 @@
-    dir,
-    timezone,
-    locale,
+    radioSchedule: { hasRadioSchedule },
```
Minor comment, would it be safer to use `pathOr` here as well - we don't validate the service config so would guard against an error here

itimirichard committed:
refactor to use pathOr

itimirichard committed:
Merge remote-tracking branch 'origin/latest' into add-radioschedules-to-frontpage

itimirichard committed:
Merge branch 'latest' into add-radioschedules-to-frontpage

PriyaKR commented:
Testing:
Looks good http://localhost:7080/arabic tested with live data across desktop browsers and devices would be covered when its on test env.
Noticed an issue on IE11 the card heights are not consistent.
![Screenshot 2020-03-13 at 14 04 44](https://user-images.githubusercontent.com/9802855/76628045-04b69200-6534-11ea-93e4-6f71d7a88d0b.png)

Commit:
Update bundle size.

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into add-radioschedules-to-frontpage

OlgaLyubin commented:
> Testing:
> Looks good http://localhost:7080/arabic tested with live data across desktop browsers and devices would be covered when its on test env.
> Noticed an issue on IE11 the card heights are not consistent.
> ![Screenshot 2020-03-13 at 14 04 44](https://user-images.githubusercontent.com/9802855/76628045-04b69200-6534-11ea-93e4-6f71d7a88d0b.png)

IE issue with different card heights is known. And it's been decided not to fix this before release.

OlgaLyubin committed:
Merge branch 'latest' into add-radioschedules-to-frontpage

Commit:
Merge branch 'add-radioschedules-to-frontpage' of github.com:bbc/simorgh into add-radioschedules-to-frontpage

PriyaKR commented:
IE is now a lower priority browser,  it’s decided not to fix this issue prior to launch.
